<a href="https://colab.research.google.com/github/LS10432/PDDS-PYD-0825-06Oct2025B/blob/main/02/Titanic_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Task 1: Loading the Data:
* Load the Titanic dataset into a Pandas DataFrame.

In [299]:
# Import the pandas library and give it an alias 'pd'
import pandas as pd

# Read the Titanic dataset and load it into a DataFrame named 'df'
df = pd.read_csv('/content/titanic.csv')

# Display the entire DataFrame in the output
display(df)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### Task 2: Initial Exploration:
* Display the first 5 rows of the dataset to get an overview.
* Provide basic information about the dataset using the info() method.

In [300]:
# Print a separator line
print("="*60)
# Print header for First 5 rows
print("First 5 rows of the DataFrame")
# Print a separator line
print("="*60)

# Print the first 5 rows of the DataFrame
display(df.head())

print("\n")

# Print a newline followed by a separator line
print("\n" + "="*60)
# Print header for Basic information
print("Basic information of the DataFrame")
# Print a separator line
print("="*60)

# Print a basic information of the DataFrame
df.info()

First 5 rows of the DataFrame


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S




Basic information of the DataFrame
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Task 3: Handling Missing Values:
* Identify and count the number of missing values in each column.
* Decide on an appropriate strategy for handling missing values in at least two columns and implement it.

In [301]:
# Code to count the number of rows in the DataFrame
num_rows = df.shape[0]
print(f"The number of rows in the DataFrame is: {num_rows}")

print("\n")

# Print a separator line
print("="*60)
# Print header for Missing values
print("Missing values before handling:")
# Print a separator line
print("="*60)

# Identify and count missing values in each column
print(df.isnull().sum())

# Handle missing values in 'Age' column by replacing missing value with the mean value
df.fillna({"Age": df["Age"].mean()}, inplace=True)

# Handle missing values in 'Cabin' column by replacing missing value with the mode value
df.fillna({"Cabin": df["Cabin"].mode()[0]}, inplace=True)

# Handle missing values in 'Embarked' column by replacing missing value with the mode value
df.fillna({"Embarked": df["Embarked"].mode()[0]}, inplace=True)

# Print a newline followed by a separator line
print("\n"+"="*60)
# Print header for Missing values
print("Missing values after handling:")
# Print a separator line
print("="*60)

# Identify and count missing values in each column
print(df.isnull().sum())

# Code to count the number of rows in the DataFrame
num_rows = df.shape[0]
print("\n")
print(f"The number of rows in the DataFrame is: {num_rows}")

The number of rows in the DataFrame is: 891


Missing values before handling:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Missing values after handling:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64


The number of rows in the DataFrame is: 891


### Task 4: Descriptive Statistics:
* Generate descriptive statistics for the numerical columns using the describe() method.

In [302]:
# Print a separator line
print("="*60)
# Print header for Descriptive statistics
print("Descriptive statistics")
# Print a separator line
print("="*60)

# Generate descriptive statistics for numerical columns
display(df.describe())

Descriptive statistics


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   13.002015    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   29.699118    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

### Task 5: Unique Values:
* List unique values in the 'Pclass' column.
* Count the occurrences of each unique value in the 'Embarked' column.

In [303]:
# Print the unique values in the 'Pclass' column
print("Unique values in 'Pclass' column:")
display(df['Pclass'].unique())

# Print a newline followed by a separator line
print("\n"+"="*60)
# Print header for Occurrences
print("Occurrences of unique values in 'Embarked' column")
# Print a separator line
print("="*60)

# Count the occurrences of each unique value in the 'Embarked' column
display(df['Embarked'].value_counts())

Unique values in 'Pclass' column:


array([3, 1, 2])


Occurrences of unique values in 'Embarked' column


Embarked
S    646
C    168
Q     77
Name: count, dtype: int64

### Task 6: Data Manipulation:
* Create a new column named 'FamilySize' that represents the sum of 'SibSp' and 'Parch'.
* Drop the 'Cabin' column from the dataset.

In [304]:
# Create a new column named 'FamilySize' that represents the sum of 'SibSp' and 'Parch'
df['FamilySize'] = df['SibSp'] + df['Parch']

# Drop the 'Cabin' column from the dataset
df = df.drop('Cabin', axis=1)

# Display the first few rows to show the changes
display(df.head())

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  FamilySize  
0      0         A/5 21171   7.2500        S           1  
1      0          PC 17599  71.2833        C           1  
2      0  STON/O2. 3101282   7.9250        S           0  
3      0            113803  53.1000        S           1  
4      0            373450   8.0500        S           0

### Task 7: Filtering and Sorting:
* Create a new DataFrame containing only passengers with an age greater than 30.
* Sort the DataFrame by 'Fare' in descending order.

In [305]:
# Create a new DataFrame containing only passengers with an age greater than 30
df_age_over_30 = df[df['Age'] > 30]

# Print a newline followed by a separator line
print("="*60)
# Print header for Occurrences
print("Passengers with an age greater than 30")
# Print a separator line
print("="*60)

# Display the first 5 rows of the filtered DataFrame
display(df_age_over_30.head())

print("\n")

# Sort the DataFrame by 'Fare' in descending order
df_sorted_by_fare = df_age_over_30.sort_values(by='Fare', ascending=False)

# Print a newline followed by a separator line
print("="*60)
# Print header for Occurrences
print("Sort by 'Fare' in descending order")
# Print a separator line
print("="*60)

# Display the first 5 rows of the sorted DataFrame
display(df_sorted_by_fare.head())

Passengers with an age greater than 30


PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
4             5         0       3   
6             7         0       1   
11           12         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   

    Parch    Ticket     Fare Embarked  FamilySize  
1       0  PC 17599  71.2833        C           1  
3       0    113803  53.1000        S           1  
4       0    373450   8.0500        S           0  
6       0     17463  51.8625        S           0  
11      0    113783  26.5500        S           0



Sort by 'Fare' in descending order


PassengerId  Survived  Pclass  \
679          680         1       1   
737          738         1       1   
258          259         1       1   
438          439         0       1   
299          300         1       1   

                                                Name     Sex   Age  SibSp  \
679               Cardeza, Mr. Thomas Drake Martinez    male  36.0      0   
737                           Lesurer, Mr. Gustave J    male  35.0      0   
258                                 Ward, Miss. Anna  female  35.0      0   
438                                Fortune, Mr. Mark    male  64.0      1   
299  Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   

     Parch    Ticket      Fare Embarked  FamilySize  
679      1  PC 17755  512.3292        C           1  
737      0  PC 17755  512.3292        C           0  
258      0  PC 17755  512.3292        C           0  
438      4     19950  263.0000        S           5  
299      1  PC 17558  247.5208        C           1

### Task 8: Data Analysis:
* Calculate the survival rate for different passenger classes ('Pclass').
* Identify and display the passenger with the highest 'Fare'.

In [306]:
# Calculate the survival rate for different passenger classes ('Pclass')
survival_rate_by_pclass = df.groupby('Pclass')['Survived'].mean()

# Print a newline followed by a separator line
print("="*60)
# Print header for Survival rate
print("Survival rate by Passenger Class:")
# Print a separator line
print("="*60)

# Display Survival rate by Passenger Class as percentages with two decimal places for better readability.
display(survival_rate_by_pclass.apply(lambda x: f'{x:.2%}'))

print("\n")

# Identify and display the passenger with the highest 'Fare'
passenger_highest_fare = df.loc[df['Fare'].idxmax()]

# Print a newline followed by a separator line
print("="*60)
# Print header for Passenger with the highest Fare
print("Passenger with the highest Fare:")
# Print a separator line
print("="*60)

# Display Passenger with the highest Fare
display(passenger_highest_fare)

Survival rate by Passenger Class:


Pclass
1    62.96%
2    47.28%
3    24.24%
Name: Survived, dtype: object



Passenger with the highest Fare:


PassengerId                 259
Survived                      1
Pclass                        1
Name           Ward, Miss. Anna
Sex                      female
Age                        35.0
SibSp                         0
Parch                         0
Ticket                 PC 17755
Fare                   512.3292
Embarked                      C
FamilySize                    0
Name: 258, dtype: object

### Conclusion:
The analysis of the Titanic dataset revealed several key findings:
- Missing values were encountered in the ‘Age’, ‘Cabin’, and ‘Embarked’ columns. These were addressed by inputting missing ‘Age’ value with the average value using the mean() function, missing ‘Cabin’ and ‘Embarked’ values with the most frequent value using the mode() function, respectively.
- The survival rate exhibited substantial disparities across passenger classes, with first-class passengers demonstrating the highest survival rate at approximately 63%, while third-class passengers exhibited the lowest at around 24%.
- The passenger with the highest fare was identified as being in first class and having survived.
- A new column named ‘FamilySize’ was created by combining ‘SibSp’ and ‘Parch’, thereby providing insights into the size of families onboard.
- The ‘Cabin’ column was dropped due to a substantial number of missing values.
- Filtering and sorting indicated that passengers exceeding the age of 30 and possessing higher fares were predominantly found in higher classes.